ca_tools/scripts/pid_controller.py

```diff
@@ -0,0 +1,238 @@
+        """
+        self._my_pid_lineal_effort = data.data
+
+    def _callback(self, data):
+        """Callback for groundtruth
+
+        Arguments:
+            data {[Odometry]} -- [Odometry msg from gts]
+        """
+        self._my_pose.x = data.pose.pose.position.x  # Update pose
+        self._my_pose.y = data.pose.pose.position.y
+        q = data.pose.pose.orientation  # get orientation quaternion
+        self._path_threshold -= 1
+        q_arr = [q.x, q.y, q.z, q.w]
+        (_,_,self._my_pose.theta) = euler_from_quaternion(q_arr)
+        if (self._path_threshold == 0):  # add the pose to path every PATH_THRESHOLD cycles
+            aux = PoseStamped()
+            aux.pose = data.pose.pose
+            aux.header.frame_id = "/map"
+            self._my_path.header.frame_id = "/map"
+            self._my_path.poses.append(aux)
+            self._my_path_publisher.publish(self._my_path)
+            self._path_threshold = PATH_THRESHOLD
+
+    def move(self):
+        if not (self._my_pose.x == None or self._my_pose.y == None):
+            """State machine
+            """
+            self._my_pid_lineal_setpoint_pub.publish(0.0)
+            self._my_pid_angular_setpoint_pub.publish(0.0)
+            self._set_goal_angle()
+            self._my_pid_lineal_state_pub.publish(self._diff_distance())
+            self._my_pid_angular_state_pub.publish(self._diff_angle())
+
+            if self._state == STATES.TURNING:
+                self._turning()
+                return
+            if self._state == STATES.FORWARD:
+                self._forward()
+                return
+            if self._state == STATES.STOP_MOVING_FORWARD:
+                self._stop_moving_forward()
+                return
+
+    def _forward(self):
+        """Makes the robot move forward
+        """
+        if self._reached_position():
+            rospy.loginfo("Reached position")
+            self._stop()
+            self._state = STATES.STOP_MOVING_FORWARD
+            return
+        aux = Twist()
+        aux.angular.z = -self._my_pid_angular_effort/PID_ANGULAR_EFFORT_ADJUST
+        aux.linear.x = -self._my_pid_lineal_effort/PID_LINEAL_EFFORT_ADJUST
+        self._my_vel_pub.publish(aux)
+
+    def _stop(self):
+        """Stops the robot
+        """
+        aux = Twist()
+        self._my_vel_pub.publish(aux)
+
+    def _diff_angle(self):
+        """Calculates the difference between the angle goal and current angle
+
+        Returns:
+            [float] -- [Difference of angles in radians, wrapped between -pi/2 and pi/2]
+        """
+        auxangle = (self._my_angle_goal-self._my_pose.theta)
+
+        while (abs(auxangle) > pi):
+            if auxangle < (-pi):
+                auxangle += (2*pi)
+            elif auxangle > pi:
+                auxangle -= (2*pi)
+        return auxangle
+
+    def _stop_moving_forward(self):
+        """Sets new goal
+        """
+        self._stop()
+        self._goal_num += 1
```

Comment:
This is not the place to increment the goal. If this the initial state, then the goal will increment before it can be reached. Also, if someone calls the function more than once, it will skip the goal without previous warning.

Reply:
Adressed